### ISSUE TITLE
SCardGetStatusChange: Race condition when attaching multiple readers at the same time

### ISSUE THREAD

**Issue by kgraefe** [LudovicRousseau/PCSC#190]
### Versions

```sh
❯ /usr/sbin/pcscd --version
pcsc-lite version 2.0.0.
Copyright (C) 1999-2002 by David Corcoran <corcoran@musclecard.com>.
Copyright (C) 2001-2022 by Ludovic Rousseau <ludovic.rousseau@free.fr>.
Copyright (C) 2003-2004 by Damien Sauveron <sauveron@labri.fr>.
Report bugs to <pcsclite-muscle@lists.infradead.org>.
Enabled features: Linux x86_64-pc-linux-gnu libsystemd serial usb libudev usbdropdir=/usr/lib/pcsc/drivers ipcdir=/run/pcscd filter configdir=/etc/reader.conf.d
MAX_READERNAME: 128, PCSCLITE_MAX_READERS_CONTEXTS: 16
❯ pcsc_scan -V
PC/SC device scanner
V 1.6.2 (c) 2001-2022, Ludovic Rousseau <ludovic.rousseau@free.fr>
```

### Platform

* Operating system or GNU/Linux distribution name and version: Ubuntu 23.10
* Smart card middleware name and version
* Smart card reader manufacturer name and reader model name: `Generic USB2.0-CRW [Smart Card Reader Interface] (20070818000000000)`
* Smart card name

### Issue

* What do you do? run `pcsc_scan -n` and attach a USB hub with two USB card readers attached
* What result do you expect? both readers should be detected
* What result do you get instead? 
  - sometimes both readers are detected in either [one](https://github.com/user-attachments/files/15785656/both_readers_detected_one_iteration.txt) or [two](https://github.com/user-attachments/files/15785660/both_readers_detected_two_iterations.txt) iterations
  - sometimes [only one reader](https://github.com/user-attachments/files/15785662/one_reader_detected.txt) is detected
  - sometimes [no reader](https://github.com/user-attachments/files/15785667/no_reader_detected.txt) is detected at all

I encountered the issue in my own program using `rust-pcsc` but could also reproduce with `pcsc_scan`. I think what happens is the following:
1. Both readers are enumerated sequentially
2. When the first reader is enumerated `\\?PnP?\Notification` fires, `SCardGetStatusChange` returns. What happens next, depends on when the second reader is enumerated
3. `pcsc_scan` calls `SCardListReaders` to get the number of readers. If the second reader is already enumerated, both will be detected.
4. `pcsc_scan` calls `SCardListReaders` with the buffer. If the second reader is now enumerated but wasn't before, the call will fail with `SCARD_E_INSUFFICIENT_BUFFER` and none will be detected.
5. `pcsc_scan` calls `SCardListReaders`. If the second reader is now enumerated but wasn't before, `\\?PnP?\Notification` does not fire and the second reader will not be detected. Otherwise it will fire at some point and go back to 3. so that both readers are detected.

In my own program I can mitigate 4. quite easily. However, how to treat 5.? The only thing I could come up with is polling but the point of `SCardGetStatusChange` is to avoid that.

### Log
see above

**Comment by LudovicRousseau** [comment]
Yes, it is a known limitation of `SCardGetStatusChange()`. You can't detect a reader connection/disconnection between 2 calls of `SCardGetStatusChange()`.

One solution would be to use `dwEventState` for the special reader `\\?PnP?\Notification` to store the number of reader events. If that number has changed between two calls of `SCardGetStatusChange()` then we return.

This field is already used by Windows to store the number of readers. I could not find the documentation for that.
See https://salsa.debian.org/rousseau/PCSC/-/issues/5

**Comment by kgraefe** [comment]
Thank you for your response. IMHO an event count would be better as the number of readers may loose events if a reader is detached and another is attached between two calls of `SCardGetStatusChange()`. However, that seems rather theoretic and I cannot oversee the Windows / macOS situation.

I read your comment `I don't think it is a good idea to implement something that would work for pcsc-lite only, and not for Windows or macOS.` so that this will not / cannot be fixed any time soon, sadly. Feel free to close the ticket if you like.

**Comment by LudovicRousseau** [comment]
I may implement the idea I presented. It will only work on GNU/Linux but at least it would be possible to have a "safe" `SCardGetStatusChange()` on this platform.

I do not expect Apple or Microsoft to do anything on their own implementation of WinSCard.

**Comment by LudovicRousseau** [comment]
I think I fixed the issue in a beta version.
See http://lists.infradead.org/pipermail/pcsclite-muscle/2024-June/001498.html

**Comment by LudovicRousseau** [comment]
Hello @kgraefe 
Have you had a chance to test my proposed code?
Does it solve the problem for you?

**Comment by kgraefe** [comment]
> Does it solve the problem for you?

yes it does! thank you :+1:

**Comment by LudovicRousseau** [comment]
Fixed in pcsc-lite 2.3.0.

See https://blog.apdu.fr/posts/2024/08/improved-scardgetstatuschange-for-pnpnotification-special-reader/
